File: frontend/src/types/finance.ts
export type AccountType = "debit" | "credit" | "investment" | "other";
export type FinanceCategoryType = "credit" | "debit" | "transfer";

export interface FinanceAccount {
  id: number;
  name: string;
  type: AccountType;
  start_date: string;
  end_date: string | null;
  archived: boolean;
  profit_eligible: boolean;
  created_at: string;
}

export interface AccountValue {
  id: number;
  account_id: number;
  year: number;
  month: number;
  amount: string;
}

export interface FinanceCategory {
  id: number;
  name: string;
  type: FinanceCategoryType;
  created_at: string;
}

export interface Merchant {
  id: number;
  name: string;
  archived: boolean;
  created_at: string;
}

export interface Transaction {
  id: number;
  date: string;
  amount: string;
  account_id: number;
  category_id: number;
  merchant_id: number | null;
  to_account_id: number | null;
  account: FinanceAccount;
  category: FinanceCategory;
  merchant: Merchant | null;
  to_account: FinanceAccount | null;
  created_at: string;
  updated_at: string;
}

export interface TransactionCreate {
  date: string;
  amount: string;
  account_id: number;
  category_id: number;
  merchant_id?: number | null;
  to_account_id?: number | null;
}

export interface TransactionUpdate {
  date?: string;
  amount?: string;
  account_id?: number;
  category_id?: number;
  merchant_id?: number;
  to_account_id?: number | null;
}

export interface AccountMonthlySummary {
  account: FinanceAccount;
  value: AccountValue | null;
  delta: string;
  reconciled: boolean;
}

export interface CategoryMonthlySummary {
  category: FinanceCategory;
  total: string;
}

export interface CategoriesByType {
  credit: CategoryMonthlySummary[];
  debit: CategoryMonthlySummary[];
  transfer: CategoryMonthlySummary[];
}

export interface MonthlySummary {
  year: number;
  month: number;
  transactions: Transaction[];
  accounts: AccountMonthlySummary[];
  categories: CategoriesByType;
  monthly_profit: string;
}

File: backend/app/schemas/finance.py
from datetime import date as Date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.finance import AccountType, FinanceCategoryType


# ── Account ───────────────────────────────────────────────────────────────────

class AccountCreate(BaseModel):
    name: str = Field(..., max_length=200)
    type: AccountType
    start_date: Date
    end_date: Date | None = None
    archived: bool = False
    profit_eligible: bool = True


class AccountUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    type: AccountType | None = None
    start_date: Date | None = None
    end_date: Date | None = None
    archived: bool | None = None
    profit_eligible: bool | None = None


class AccountRead(BaseModel):
    id: int
    name: str
    type: AccountType
    start_date: Date
    end_date: Date | None
    archived: bool
    profit_eligible: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ── AccountValue ──────────────────────────────────────────────────────────────

class AccountValueUpsert(BaseModel):
    amount: Decimal = Field(..., decimal_places=2)


class AccountValueRead(BaseModel):
    id: int
    account_id: int
    year: int
    month: int
    amount: Decimal

    model_config = {"from_attributes": True}


# ── FinanceCategory ───────────────────────────────────────────────────────────

class FinanceCategoryCreate(BaseModel):
    name: str = Field(..., max_length=100)
    type: FinanceCategoryType


class FinanceCategoryUpdate(BaseModel):
    name: str | None = Field(None, max_length=100)
    type: FinanceCategoryType | None = None


class FinanceCategoryRead(BaseModel):
    id: int
    name: str
    type: FinanceCategoryType
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Merchant ──────────────────────────────────────────────────────────────────

class MerchantCreate(BaseModel):
    name: str = Field(..., max_length=200)
    archived: bool = False


class MerchantUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    archived: bool | None = None


class MerchantRead(BaseModel):
    id: int
    name: str
    archived: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Transaction ───────────────────────────────────────────────────────────────

class TransactionCreate(BaseModel):
    date: Date
    amount: Decimal = Field(..., decimal_places=2)
    account_id: int | None = None
    category_id: int
    merchant_id: int | None = None
    to_account_id: int | None = None


class TransactionUpdate(BaseModel):
    date: Date | None = None
    amount: Decimal | None = Field(None, decimal_places=2)
    account_id: int | None = None
    category_id: int | None = None
    merchant_id: int | None = None
    to_account_id: int | None = None


class TransactionRead(BaseModel):
    id: int
    date: Date
    amount: Decimal
    description: str | None
    account_id: int | None
    category_id: int
    merchant_id: int | None
    to_account_id: int | None
    account: AccountRead | None
    category: FinanceCategoryRead
    merchant: MerchantRead | None
    to_account: AccountRead | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Monthly Summary ───────────────────────────────────────────────────────────

class AccountMonthlySummary(BaseModel):
    account: AccountRead
    value: AccountValueRead | None
    delta: Decimal
    reconciled: bool

    model_config = {"from_attributes": True}


class CategoryMonthlySummary(BaseModel):
    category: FinanceCategoryRead
    total: Decimal

    model_config = {"from_attributes": True}


class CategoriesByType(BaseModel):
    credit: list[CategoryMonthlySummary]
    debit: list[CategoryMonthlySummary]
    transfer: list[CategoryMonthlySummary]


class MonthlySummary(BaseModel):
    year: int
    month: int
    transactions: list[TransactionRead]
    accounts: list[AccountMonthlySummary]
    categories: CategoriesByType
    monthly_profit: Decimal
